source/nsdynmemLIB/nsdynmemLIB.c

```diff
@@ -290,178 +238,49 @@ void *ns_dyn_mem_alloc(int16_t alloc_size)
-            }
-        }
-        platform_exit_critical();
-    }
-    return retval;
-
-#else
-
-    void *retval = 0;
-    if (alloc_size) {
-        platform_enter_critical();
-        retval = malloc(alloc_size);
-        platform_exit_critical();
-    }
-    return retval;
-#endif
+    return ns_dyn_mem_internal_alloc(alloc_size, 1);
 }
 
 #ifndef STANDARD_MALLOC
-static void ns_free_and_merge_next_sector(int *cur_sector, int size)
+static void ns_free_and_merge_with_adjacent_blocks(int *cur_block, int size)
 {
-    int *end_sector;
-    int next_len;
-    end_sector = cur_sector;
-    end_sector += size + 1;
-    if (end_sector != heap_main_end) {
-        end_sector++;
-        next_len = *end_sector;
-
-        if (next_len > 0) {
-            //Convert End of len
-            end_sector--;
-        } else {
-            //Merge First to second
-            //Set First End
-            //Skip Data And sector start of len
-            //debug("Free Next\r\n");
-            next_len = -(next_len);
-            end_sector += (next_len + 1);
-            size += (next_len + 2);
+    // Theory of operation: Block is always in form | Len | Data | Len |
+    // So we need to check length of previous (if current not heap start)
+    // and next (if current not heap end) blocks. Negative length means
+    // free memory so we can merge freed block with those.
+
+    int *start = cur_block;
+    int *end = cur_block + size + 1;
+    int total_size = size;
+
+    if (cur_block != heap_main) {
+        cur_block--;
+        if (*cur_block < 0) {
+            total_size += (2 - *cur_block);
+            start -= (2 - *cur_block);
         }
-
+        cur_block++;
     }
-    //Set new sector start & End length
-    *cur_sector = -(size);
-    *end_sector = -(size);
-}
 
-static void ns_merge_prev_sector(int *cur_sector)
-{
-    int *prev_sector = 0;
-    int prev_len, cur_len;
-
-    if (cur_sector != heap_main) {
-        prev_sector = cur_sector;
-        prev_sector--;
-        prev_len = *prev_sector;
-
-        if (prev_len < 0) {
-            //Merge First to second
-            //Set First End
-            //Skip Data And sector start of len
-            //debug("Merge Prev\r\n");
-            prev_len = -(prev_len);
-            prev_sector -= (prev_len + 1); // SET PREV Start
-
-            cur_len = *cur_sector;
-            cur_len = -cur_len; //Invert
-            cur_sector += (cur_len + 1); //New End pointer
-            prev_len += (cur_len + 2); // SET NEW LEN
-        } else {
-            prev_sector = 0;
+    if (end != heap_main_end) {
+        end++;
+        if (*end < 0) {
+            total_size += (2 - *end);
+            end += (1 - *end);
+        }else{
+            end--;
         }
-
-    }
-    //Set new sector start & End length
-    if (prev_sector) {
-        *prev_sector = -(prev_len);
-        *cur_sector = -(prev_len);
     }
+    *cur_block = -total_size;
```

Comment:
Actually, shouldn't this be -data_size (the passed-in-size), technically? Doesn't really matter, but might be less confusing when dissecting crash dumps.

And would make sense to do this FIRST, before starting the merge. (That would also avoid simplify the compiler's job if making the above change, as it wouldn't need to remember data_size).

Maybe it shouldn't even be part of this call - it should be done outside this call, and this should just be a merge. (Which in turn would not bother with a separate size parameter, it would just load from the pointer).